### PR TITLE
Window size review #508

### LINKED ISSUE #508: Window size meaning
* add docs / etc for meaning of window (about it is half of pattern window in fact)
or
* add changes that make this half-window full size

### PR DESCRIPTION
**Pull request by amper43**
fixes #508 

changes:
- change sence of window size from half window to full window
- added migration for existing window sizes

### CODE REVIEW

File: analytics/analytics/detectors/pattern_detector.py [hunk 1]
```diff
@@ -36,8 +36,8 @@ def resolve_model_by_pattern(pattern: str) -> models.Model:
 class PatternDetector(Detector):
 
     MIN_BUCKET_SIZE = 150
-    BUCKET_WINDOW_SIZE_FACTOR = 5
-    DEFAULT_WINDOW_SIZE = 1
+    BUCKET_WINDOW_SIZE_FACTOR = 3
+    DEFAULT_WINDOW_SIZE = 3
```
Comment (rozetko) [inline analytics/analytics/detectors/pattern_detector.py]: Why is it 3, not 2?

File: analytics/analytics/models/general_model.py [hunk 2]
```diff
@@ -85,18 +86,19 @@ def do_detect(self, dataframe: pd.DataFrame, id: AnalyticUnitId) -> List[int]:
         if pat_data.count(0) == len(pat_data):
             raise ValueError('Labeled patterns must not be empty')
 
+        half_window_size = self.state.get('WINDOW_SIZE', 0) // 2
```
Comment (rozetko) [inline analytics/analytics/models/general_model.py]: Let's add a `TODO` comment about removing `half_window_size`

File: analytics/analytics/models/general_model.py [hunk 3]
```diff
@@ -85,18 +86,19 @@ def do_detect(self, dataframe: pd.DataFrame, id: AnalyticUnitId) -> List[int]:
...
+        return [(item, item + window_size) for item in filtered]
 
     def __filter_detection(self, segments:  Generator[int, None, None], data: pd.Series) -> Generator[int, None, None]:
         if not self.state.get('pattern_center'):
             return []
-        window_size = self.state.get('WINDOW_SIZE', 0)
+        half_window_size = self.state.get('WINDOW_SIZE', 0) // 2
         pattern_model = self.state.get('pattern_model', [])
         for ind, val in segments:
             watch_data = data[ind - window_size: ind + window_size + 1]
```
Comment (rozetko) [inline analytics/analytics/models/general_model.py]: You use common `window_size` here but define `half_window_size` earlier
Seems like `half_window_size` should be used here

File: analytics/analytics/models/peak_model.py [hunk 4]
```diff
@@ -89,8 +89,8 @@ def do_fit(self, dataframe: pd.DataFrame, labeled_segments: list, deleted_segmen
     def do_detect(self, dataframe: pd.DataFrame, id: str):
         data = utils.cut_dataframe(dataframe)
         data = data['value']
-        window_size = int(len(data)/SMOOTHING_COEFF) #test ws on flat data
```
Comment (rozetko) [inline analytics/analytics/models/peak_model.py]: why did you remove it?

File: analytics/analytics/models/trough_model.py [hunk 5]
```diff
@@ -89,8 +89,8 @@ def do_fit(self, dataframe: pd.DataFrame, labeled_segments: list, deleted_segmen
     def do_detect(self, dataframe: pd.DataFrame, id: str):
         data = utils.cut_dataframe(dataframe)
         data = data['value']
-        window_size = int(len(data)/SMOOTHING_COEFF) #test ws on flat data
```
Comment (rozetko) [inline analytics/analytics/models/trough_model.py]: same here

File: server/src/index.ts [hunk 6]
```diff
@@ -10,6 +10,7 @@ import * as ProcessService from './services/process_service';
 import { HASTIC_PORT, PACKAGE_VERSION, GIT_INFO, ZMQ_CONNECTION_STRING, HASTIC_INSTANCE_NAME } from './config';
 
 import { convertPanelUrlToPanelId } from './migrations/0.3.2-beta';
+import { convertHalfWinowSize } from './migrations/window_size';
```
Comment (rozetko) [inline server/src/index.ts]: Typo: `convertHalfWinowSize` -> `convertHalfWindowSize`

File: server/src/migrations/window_size.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
+import { Collection, makeDBQ } from '../services/data_service';
+import * as AnalyticUnitCache from '../models/analytic_unit_cache_model';
+import * as AnalyticUnit from '../models/analytic_unit_model';
+
+const db = makeDBQ(Collection.DB_VERSION);
+const DB_INFO_ID = '0';
+const REVISION_FOR_APPLY = 1
```
Comment (rozetko) [inline server/src/migrations/window_size.ts]:
```suggestion
const REVISION_FOR_APPLY = 1;
```

File: server/src/migrations/window_size.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  _id: string,
+  revision: number
+}
+
+
+export async function convertHalfWinowSize() {
+  const db_info: DbInfo = await db.findOne(DB_INFO_ID);
+
+  if(db_info === null) {
+    await db.insertOne({_id: DB_INFO_ID, revision: 0});
```
Comment (rozetko) [inline server/src/migrations/window_size.ts]:
```suggestion
    await db.insertOne({ _id: DB_INFO_ID, revision: 0 });
```

File: server/src/models/analytic_unit_cache_model.ts [hunk 9]
```diff
@@ -38,6 +38,13 @@ export async function findById(id: AnalyticUnitId): Promise<AnalyticUnitCache> {
   return AnalyticUnitCache.fromObject(obj);
 }
 
+export async function getAllCaches(): Promise<AnalyticUnitCache[]> {
+  const caches = await db.findMany({
+    id: { $gt: '' }
```
Comment (rozetko) [inline server/src/models/analytic_unit_cache_model.ts]: What does `greater than ''` mean?

File: server/src/index.ts [hunk 10]
```diff
@@ -19,6 +20,7 @@ init();
 
 async function init() {
   await convertPanelUrlToPanelId();
+  //await convertHalfWinowSize();
```
Comment (rozetko) [inline server/src/index.ts]: btw, why is it commented?

File: analytics/analytics/models/model.py [hunk 11]
```diff
@@ -114,8 +114,8 @@ def fit(self, dataframe: pd.DataFrame, segments: list, id: AnalyticUnitId, cache
 
         assert len(labeled) > 0, f'labeled list empty, skip fitting for {id}'
 
-        if self.state.get('WINDOW_SIZE') == 0:
-            self.state['WINDOW_SIZE'] = math.ceil(max_length / 2) if max_length else 0
+        self.state['WINDOW_SIZE'] = self.state.get('WINDOW_SIZE', max_length)
```
Comment (rozetko) [inline analytics/analytics/models/model.py]: `if self.state.get('WINDOW_SIZE') == 0:`

**Comment by jonyrock** [comment]
it's to old. Please create a new PR for this.
